File: leetcode/34.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] result = {-1, -1};

        int leftIndex = findIndex(nums, target, true);
        if (leftIndex == -1) {
            return result; 
        }

        int rightIndex = findIndex(nums, target, false);

        result[0] = leftIndex;
        result[1] = rightIndex;

        return result;
    }

    private int findIndex(int[] nums, int target, boolean isLeft) {
        int index = -1;
        int start = 0;
        int end = nums.length - 1;

        while (start <= end) {
            int mid = start + (end - start) / 2;

            if (nums[mid] == target) {
                index = mid;
                if (isLeft) {
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            } else if (nums[mid] < target) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }

        return index;
    }
}

File: code/34.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)
        low = 0; high = n - 1
        ans = [-1, -1]
        while (low <= high):
            mid = (low + high)//2
            if target < nums[mid]:
                high = mid - 1
            elif (target > nums[mid]):
                low = mid + 1
            else:
                if (mid == 0 or nums[mid] != nums[mid - 1]):
                    ans[0] = mid
                    break
                else: high = mid - 1
        
        low = 0; high = n - 1
        while (low <= high):
            mid = (low + high)//2
            if (target < nums[mid]):
                high = mid - 1
            elif (target > nums[mid]):
                low = mid + 1
            else:
                if (mid == n - 1 or nums[mid] != nums[mid + 1]):
                    ans[1] = mid
                    break
                else: low = mid + 1

        return ans
                